### PR TITLE
Rework attribute analyzers to symbol analyzers

### PR DESCRIPTION
**Pull request by Dreamescaper**
Contributes to #270 .

The following analyzers are reworked as symbol analyzers:

- TestCaseUsageAnalyzer
- TestMethodUsage
- ParallelizableUsageAnalyzer 
- NonTestMethodAccessibilityLevelAnalyzer 
- TestMethodAccessibilityLevelAnalyzer 

The easiest would be to review commit-by-commit, ignoring spaces.

### CODE REVIEW

File: src/nunit.analyzers.tests/TestMethodUsage/TestMethodUsageAnalyzerTests.cs [hunk 1]
```diff
@@ -510,6 +510,7 @@ public void AnalyzeWhenSimpleTestMethodHasParameters()
 
             var testCode = TestUtility.WrapMethodInClassNamespaceAndAddUsings(@"
     [↓Test]
+    [Category(""Test"")]
```
Comment (Dreamescaper) [inline src/nunit.analyzers.tests/TestMethodUsage/TestMethodUsageAnalyzerTests.cs:513]: Previously there was a bug that diagnostic was added to non-test attributes in this case.

File: src/nunit.analyzers/ParallelizableUsage/ParallelizableUsageAnalyzer.cs [hunk 2]
```diff
@@ -46,70 +43,85 @@ public sealed class ParallelizableUsageAnalyzer : DiagnosticAnalyzer
         public override void Initialize(AnalysisContext context)
         {
             context.EnableConcurrentExecution();
-
-            context.RegisterSyntaxNodeAction(ParallelizableUsageAnalyzer.AnalyzeAttribute, SyntaxKind.Attribute);
+            context.RegisterCompilationAction(AnalyzeCompilation);
```
Comment (Dreamescaper) [inline src/nunit.analyzers/ParallelizableUsage/ParallelizableUsageAnalyzer.cs:46]: additionally registered Complilation action to analyze assembly attributes

File: src/nunit.analyzers.tests/Extensions/AttributeArgumentTypedConstantExtensionsTests.cs [hunk 3]
```diff
@@ -23,8 +23,6 @@ static IEnumerable<TestCaseData> GetTestData()
                 SetName("CanAssignToWhenArgumentIsNullAndTargetIsValueType");
             yield return new TestCaseData("\"x\"", "object", "string", Is.True).
                 SetName("CanAssignToWhenArgumentIsNotNullableAndAssignable");
-            yield return new TestCaseData("3", "int", "int?", Is.True).
```
Comment (Dreamescaper) [inline src/nunit.analyzers.tests/Extensions/AttributeArgumentTypedConstantExtensionsTests.cs:26]: this case leads to invalid code - `int?` cannot be a attribute parameter.

**Comment by mikkelbu** [comment]
I just tried the different versions against NodaTime and got results similar to yours. I think this is fine for now to close #270 and then we can later see if we can do some more (if it becomes a problem). Perhaps by profiling the individual diagnostics.

<details>
<summary>lastest release</summary>
6.976  100   nunit.analyzers, Version=0.4.0.0, Culture=neutral, PublicKeyToken=null
    6.976  100   nunit.analyzers, Version=0.4.0.0, Culture=neutral, PublicKeyToken=null
    1.674   23      NUnit.Analyzers.TestCaseUsage.TestCaseUsageAnalyzer
    1.379   19      NUnit.Analyzers.ConstraintUsage.EqualConstraintUsageAnalyzer
    0.593    8      NUnit.Analyzers.TestCaseSourceUsage.TestCaseSourceUsesStringAnalyzer
    0.561    8      NUnit.Analyzers.TestCaseUsage.TestMethodUsageAnalyzer
    0.470    6      NUnit.Analyzers.ConstActualValueUsage.ConstActualValueUsageAnalyzer
    0.321    4      NUnit.Analyzers.ClassicModelAssertUsage.ClassicModelAssertUsageAnalyzer
    0.268    3      NUnit.Analyzers.EqualToIncompatibleTypes.EqualToIncompatibleTypesAnalyzer
    0.230    3      NUnit.Analyzers.TestMethodAccessibilityLevel.TestMethodAccessibilityLevelAnalyzer
    0.202    2      NUnit.Analyzers.ConstraintUsage.SomeItemsConstraintUsageAnalyzer
    0.169    2      NUnit.Analyzers.StringConstraintWrongActualType.StringConstraintWrongActualTypeAnalyzer
    0.161    2      NUnit.Analyzers.ParallelizableUsage.ParallelizableUsageAnalyzer
    0.158    2      NUnit.Analyzers.MissingProperty.MissingPropertyAnalyzer
    0.151    2      NUnit.Analyzers.ContainsConstraintWrongActualType.ContainsConstraintWrongActualTypeAnalyzer
    0.141    2      NUnit.Analyzers.SameActualExpectedValue.SameActualExpectedValueAnalyzer
    0.123    1      NUnit.Analyzers.SameAsIncompatibleTypes.SameAsIncompatibleTypesAnalyzer
    0.102    1      NUnit.Analyzers.ConstraintUsage.StringConstraintUsageAnalyzer
    0.083    1      NUnit.Analyzers.IgnoreCaseUsage.IgnoreCaseUsageAnalyzer
    0.071    1      NUnit.Analyzers.NullConstraintUsage.NullConstraintUsageAnalyzer
    0.069   <1      NUnit.Analyzers.SomeItemsIncompatibleTypes.SomeItemsIncompatibleTypesAnalyzer
    0.051   <1      NUnit.Analyzers.ValueSourceUsage.ValueSourceUsageAnalyzer
</details>

<details>
<summary>master branch</summary>
   1.621   99   nunit.analyzers, Version=0.5.0.554, Culture=neutral, PublicKeyToken=null
       1.621   99   nunit.analyzers, Version=0.5.0.554, Culture=neutral, PublicKeyToken=null
       0.572   35      NUnit.Analyzers.TestCaseUsage.TestCaseUsageAnalyzer
       0.183   11      NUnit.Analyzers.EqualToIncompatibleTypes.EqualToIncompatibleTypesAnalyzer
       0.176   10      NUnit.Analyzers.TestCaseUsage.TestMethodUsageAnalyzer
       0.168   10      NUnit.Analyzers.TestCaseSourceUsage.TestCaseSourceUsesStringAnalyzer
       0.097    6      NUnit.Analyzers.TestMethodAccessibilityLevel.TestMethodAccessibilityLevelAnalyzer
       0.075    4      NUnit.Analyzers.ClassicModelAssertUsage.ClassicModelAssertUsageAnalyzer
       0.053    3      NUnit.Analyzers.MissingProperty.MissingPropertyAnalyzer
       0.046    2      NUnit.Analyzers.StringConstraintWrongActualType.StringConstraintWrongActualTypeAnalyzer
       0.039    2      NUnit.Analyzers.SameActualExpectedValue.SameActualExpectedValueAnalyzer
       0.033    2      NUnit.Analyzers.ConstraintUsage.EqualConstraintUsageAnalyzer
       0.030    1      NUnit.Analyzers.ValueSourceUsage.ValueSourceUsageAnalyzer
       0.029    1      NUnit.Analyzers.ParallelizableUsage.ParallelizableUsageAnalyzer
       0.026    1      NUnit.Analyzers.ConstraintUsage.SomeItemsConstraintUsageAnalyzer
       0.018    1      NUnit.Analyzers.ContainsConstraintWrongActualType.ContainsConstraintWrongActualTypeAnalyzer
       0.013   <1      NUnit.Analyzers.ComparableTypes.ComparableTypesAnalyzer
       0.012   <1      NUnit.Analyzers.ConstActualValueUsage.ConstActualValueUsageAnalyzer
       0.010   <1      NUnit.Analyzers.SameAsIncompatibleTypes.SameAsIncompatibleTypesAnalyzer
       0.009   <1      NUnit.Analyzers.ConstraintUsage.StringConstraintUsageAnalyzer
       0.009   <1      NUnit.Analyzers.SameAsOnValueTypes.SameAsOnValueTypesAnalyzer
       0.008   <1      NUnit.Analyzers.SomeItemsIncompatibleTypes.SomeItemsIncompatibleTypesAnalyzer
       0.008   <1      NUnit.Analyzers.NullConstraintUsage.NullConstraintUsageAnalyzer
       0.007   <1      NUnit.Analyzers.IgnoreCaseUsage.IgnoreCaseUsageAnalyzer
      <0.001   <1      NUnit.Analyzers.NonTestMethodAccessibilityLevel.NonTestMethodAccessibilityLevelAnalyzer
</details>

<details>
<summary>this PR branch</summary>
   1.208  100   nunit.analyzers, Version=0.5.0.552, Culture=neutral, PublicKeyToken=null
       1.208  100   nunit.analyzers, Version=0.5.0.552, Culture=neutral, PublicKeyToken=null
       0.534   44      NUnit.Analyzers.TestCaseSourceUsage.TestCaseSourceUsesStringAnalyzer
       0.136   11      NUnit.Analyzers.ParallelizableUsage.ParallelizableUsageAnalyzer
       0.095    7      NUnit.Analyzers.TestCaseUsage.TestCaseUsageAnalyzer
       0.087    7      NUnit.Analyzers.ClassicModelAssertUsage.ClassicModelAssertUsageAnalyzer
       0.068    5      NUnit.Analyzers.ConstraintUsage.SomeItemsConstraintUsageAnalyzer
       0.052    4      NUnit.Analyzers.TestMethodUsage.TestMethodUsageAnalyzer
       0.036    3      NUnit.Analyzers.ValueSourceUsage.ValueSourceUsageAnalyzer
       0.028    2      NUnit.Analyzers.SameActualExpectedValue.SameActualExpectedValueAnalyzer
       0.027    2      NUnit.Analyzers.StringConstraintWrongActualType.StringConstraintWrongActualTypeAnalyzer
       0.024    1      NUnit.Analyzers.EqualToIncompatibleTypes.EqualToIncompatibleTypesAnalyzer
       0.023    1      NUnit.Analyzers.ConstraintUsage.EqualConstraintUsageAnalyzer
       0.019    1      NUnit.Analyzers.ConstraintUsage.StringConstraintUsageAnalyzer
       0.011   <1      NUnit.Analyzers.ConstActualValueUsage.ConstActualValueUsageAnalyzer
       0.010   <1      NUnit.Analyzers.MissingProperty.MissingPropertyAnalyzer
       0.010   <1      NUnit.Analyzers.IgnoreCaseUsage.IgnoreCaseUsageAnalyzer
       0.009   <1      NUnit.Analyzers.SameAsIncompatibleTypes.SameAsIncompatibleTypesAnalyzer
       0.009   <1      NUnit.Analyzers.SomeItemsIncompatibleTypes.SomeItemsIncompatibleTypesAnalyzer
       0.009   <1      NUnit.Analyzers.SameAsOnValueTypes.SameAsOnValueTypesAnalyzer
       0.008   <1      NUnit.Analyzers.TestMethodAccessibilityLevel.TestMethodAccessibilityLevelAnalyzer
       0.006   <1      NUnit.Analyzers.ContainsConstraintWrongActualType.ContainsConstraintWrongActualTypeAnalyzer
       0.006   <1      NUnit.Analyzers.NullConstraintUsage.NullConstraintUsageAnalyzer
      <0.001   <1      NUnit.Analyzers.NonTestMethodAccessibilityLevel.NonTestMethodAccessibilityLevelAnalyzer
</details>

**Pull request merged**